Port fuse_linear from pytorch/tvm

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25974 [quant][graphmode] use whitelist for selecting observed values
* #25625 [quant][graphmode] Fold quantize op into module
* #25624 [quant][graphmode] Add fusion for quantized linear
* **#25623 Port fuse_linear from pytorch/tvm**
* #25505 [quant][graphmode] Support quantizing any methods called

Summary:
Port over fuse_linear pass from pytorch/tvm project, we'll need this
in backend specific quantization pass to match aten::linear and swap
it with quantized linear

Test Plan:
python test/test_jit.py 'TestJit.test_fuse_linear'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208890](https://our.internmc.facebook.com/intern/diff/D17208890)

jerryzh168 committed:
Port fuse_linear from pytorch/tvm

Summary:
Port over fuse_linear pass from pytorch/tvm project, we'll need this
in backend specific quantization pass to match aten::linear and swap
it with quantized linear

Test Plan:
python test/test_jit.py 'TestJit.test_fuse_linear'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "Port fuse_linear from pytorch/tvm"


Summary:
Port over fuse_linear pass from pytorch/tvm project, we'll need this
in backend specific quantization pass to match aten::linear and swap
it with quantized linear

Test Plan:
python test/test_jit.py 'TestJit.test_fuse_linear'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "Port fuse_linear from pytorch/tvm"


Summary:
Port over fuse_linear pass from pytorch/tvm project, we'll need this
in backend specific quantization pass to match aten::linear and swap
it with quantized linear

Test Plan:
python test/test_jit.py 'TestJit.test_fuse_linear'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "Port fuse_linear from pytorch/tvm"


Summary:
Port over fuse_linear pass from pytorch/tvm project, we'll need this
in backend specific quantization pass to match aten::linear and swap
it with quantized linear

Test Plan:
python test/test_jit.py 'TestJit.test_fuse_linear'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208890](https://our.internmc.facebook.com/intern/diff/D17208890)

jianyuh commented:
Just double check, this is just porting the code from pytorch/tvm, not adding a dependency to pytorch/tvm, right?

jerryzh168 commented:
Yes that is correct
On Sep 6, 2019, 13:00 -0700, Jianyu Huang <notifications@github.com>, wrote:
> Just double check, this is just porting the code from pytorch/tvm, not adding a dependency to pytorch/tvm, right?
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

nishantpdce commented on torch/csrc/jit/passes/fuse_linear.h:
```diff
@@ -0,0 +1,17 @@
+namespace torch {
+namespace jit {
+
+/** \brief Match the at::linear pattern and fuse it into a sigle at::linear
```
minor typo sigle

nishantpdce commented:
LGTM

jerryzh168 committed:
Update on "Port fuse_linear from pytorch/tvm"


Summary:
Port over fuse_linear pass from pytorch/tvm project, we'll need this
in backend specific quantization pass to match aten::linear and swap
it with quantized linear

Test Plan:
python test/test_jit.py 'TestJit.test_fuse_linear'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208890](https://our.internmc.facebook.com/intern/diff/D17208890)

jerryzh168 committed:
Update on "Port fuse_linear from pytorch/tvm"


Summary:
Port over fuse_linear pass from pytorch/tvm project, we'll need this
in backend specific quantization pass to match aten::linear and swap
it with quantized linear

Test Plan:
python test/test_jit.py 'TestJit.test_fuse_linear'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208890](https://our.internmc.facebook.com/intern/diff/D17208890)

jerryzh168 committed:
Update on "Port fuse_linear from pytorch/tvm"


Summary:
Port over fuse_linear pass from pytorch/tvm project, we'll need this
in backend specific quantization pass to match aten::linear and swap
it with quantized linear

Test Plan:
python test/test_jit.py 'TestJit.test_fuse_linear'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208890](https://our.internmc.facebook.com/intern/diff/D17208890)

jerryzh168 committed:
Update on "Port fuse_linear from pytorch/tvm"


Summary:
Port over fuse_linear pass from pytorch/tvm project, we'll need this
in backend specific quantization pass to match aten::linear and swap
it with quantized linear

Test Plan:
python test/test_jit.py 'TestJit.test_fuse_linear'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208890](https://our.internmc.facebook.com/intern/diff/D17208890)

jerryzh168 committed:
Update on "Port fuse_linear from pytorch/tvm"


Summary:
Port over fuse_linear pass from pytorch/tvm project, we'll need this
in backend specific quantization pass to match aten::linear and swap
it with quantized linear

Test Plan:
python test/test_jit.py 'TestJit.test_fuse_linear'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208890](https://our.internmc.facebook.com/intern/diff/D17208890)

ZolotukhinM commented on torch/csrc/jit/passes/fuse_linear.cpp:
```diff
@@ -0,0 +1,52 @@
+
+    // replace matmul + add pattern to linear
+    SubgraphRewriter matmuladd_to_linear;
+    matmuladd_to_linear.RegisterRewritePattern(matmul_add_pattern, fused_linear);
```
You could register all of the patterns together and then call `runOnGraph` once.

ZolotukhinM commented on torch/csrc/jit/passes/fuse_linear.cpp:
```diff
@@ -0,0 +1,52 @@
+        return (%output))IR";
+    std::string fused_linear_bias_none = R"IR(
+    graph(%input, %weight):
+        %bias: Tensor? = prim::Constant()
```
Are you sure this one would run correctly after the transformation?

ZolotukhinM reviewed (approved):
Overall looks good, I have a couple of comments (some here, some inline):

 - When you add a new file, you also need to add it to `tools/build_variables.py`, otherwise internal builds will fail.
 - [Nit] The commit message mentions tvm for no reason - it's not like you ported a pass from TVM, you just implemented a similar pass in PyTorch. It's good to mention the source of origin/inspiration, but I suspect there was not much "porting" involved in this PR.

jerryzh168 committed:
Update on "Port fuse_linear from pytorch/tvm"


Summary:
Port over fuse_linear pass from pytorch/tvm project, we'll need this
in backend specific quantization pass to match aten::linear and swap
it with quantized linear

Test Plan:
python test/test_jit.py 'TestJit.test_fuse_linear'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208890](https://our.internmc.facebook.com/intern/diff/D17208890)

jerryzh168 committed:
Update on "Port fuse_linear from pytorch/tvm"


Summary:
Port over fuse_linear pass from pytorch/tvm project, we'll need this
in backend specific quantization pass to match aten::linear and swap
it with quantized linear

Test Plan:
python test/test_jit.py 'TestJit.test_fuse_linear'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208890](https://our.internmc.facebook.com/intern/diff/D17208890)

jerryzh168 committed:
Update on "Port fuse_linear from pytorch/tvm"


Summary:
Port over fuse_linear pass from pytorch/tvm project, we'll need this
in backend specific quantization pass to match aten::linear and swap
it with quantized linear

Test Plan:
python test/test_jit.py 'TestJit.test_fuse_linear'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208890](https://our.internmc.facebook.com/intern/diff/D17208890)

jerryzh168 committed:
Update on "Port fuse_linear from pytorch/tvm"


Summary:
Port over fuse_linear pass from pytorch/tvm project, we'll need this
in backend specific quantization pass to match aten::linear and swap
it with quantized linear

Test Plan:
python test/test_jit.py 'TestJit.test_fuse_linear'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208890](https://our.internmc.facebook.com/intern/diff/D17208890)
